src/ProtoBuf.FSharp/CodeGen.fs

```diff
@@ -66,99 +119,381 @@ let private emitFactory (resultType : Type) (zeroValuesPerField: ZeroValues.Fiel
     gen.Emit(OpCodes.Ret)
     factoryMethod :> MethodInfo
 
+
+let private getGenericArgs (t : Type) =
+    if t.IsGenericTypeDefinition then
+        t.GetGenericArguments() |> ValueSome
+    else ValueNone
+
+let private defineGenericArgs (args : ValueOption<Type[]>) (tb : TypeBuilder) =
+    args |> ValueOption.iter (fun args -> tb.DefineGenericParameters [| for arg in args -> arg.Name |] |> ignore)
+
+let private substituteGenericArgs args (t : Type) =
+    match args with
+    | ValueNone -> t
+    | ValueSome args -> t.MakeGenericType args
+
+let private emitSurrogateContent (tb : TypeBuilder) (targetType : Type) (targetFields : PropertyInfo[]) (targetGenerate : MethodBase) (isVirtual : bool) (baseConstructor : ConstructorInfo) =
+    let fields = [| for fi in targetFields -> struct (fi, tb.DefineFieldForProtobuf(fi)) |]
+    let constructor =
+        let paramType = if targetType.IsValueType then targetType.MakeByRefType() else targetType
+        tb.DefineConstructor(MethodAttributes.Public, CallingConventions.Standard, [| paramType |])
+    begin
+        let gen = constructor.GetILGenerator()
+        if not tb.IsValueType then
+            gen.Emit(OpCodes.Ldarg_0)
```

Comment:
English or C#/F# equivalent code as comments please at each step. With enough time I can kinda grok what the end surrogate looks like but making it quicker helps.